tokio-net/src/driver/sharded_slab/mod.rs

```diff
@@ -0,0 +1,317 @@
+    /// number of shards has been reached.
+    pub(crate) fn insert(&self, value: T) -> Option<usize> {
+        let tid = Tid::current();
+        self.shards[tid.as_usize()]
+            .with_mut(|shard| unsafe {
+                // we are guaranteed to only mutate the shard while on its thread.
+                (*shard).insert(value)
+            })
+            .map(|idx| tid.pack(idx))
+    }
+
+    /// Removes the value associated with the given key from the slab, returning
+    /// it.
+    ///
+    /// If the slab does not contain a value for that key, `None` is returned
+    /// instead.
+    pub(crate) fn remove(&self, idx: usize) -> Option<T> {
+        let tid = Tid::from_packed(idx);
+        if tid.is_current() {
+            self.shards[tid.as_usize()].with_mut(|shard| unsafe {
+                // only called if this is the current shard
+                (*shard).remove_local(idx)
+            })
+        } else {
+            self.shards[tid.as_usize()].with(|shard| unsafe { (*shard).remove_remote(idx) })
+        }
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the slab does not contain a value for the given key, `None` is
+    /// returned instead.
+    pub(crate) fn get(&self, key: usize) -> Option<&T> {
+        let tid = Tid::from_packed(key);
+        self.shards
+            .get(tid.as_usize())?
+            .with(|shard| unsafe { (*shard).get(key) })
+    }
+
+    /// Returns the number of items currently stored in the slab.
+    pub(crate) fn len(&self) -> usize {
+        self.shards
+            .iter()
+            .map(|shard| shard.with(|shard| unsafe { (*shard).len() }))
+            .sum()
+    }
+
+    /// Returns an iterator over all the items in the slab.
+    pub(crate) fn unique_iter(&mut self) -> iter::UniqueIter<'_, T> {
+        let mut shards = self.shards.iter_mut();
+        let shard = shards.next().expect("must be at least 1 shard");
+        let mut pages = shard.with(|shard| unsafe { (*shard).iter() });
+        let slots = pages.next().expect("must be at least 1 page").iter();
+        iter::UniqueIter {
+            shards,
+            slots,
+            pages,
+        }
+    }
+}
+
+impl<T> Shard<T> {
+    fn new(_tid: usize) -> Self {
+        Self {
+            #[cfg(debug_assertions)]
+            tid: _tid,
+            sz: page::INITIAL_SIZE,
+            len: AtomicUsize::new(0),
+            pages: vec![Page::new(page::INITIAL_SIZE, 0)],
+        }
+    }
+
+    fn insert(&mut self, value: T) -> Option<usize> {
+        #[cfg(debug_assertions)]
+        debug_assert_eq!(Tid::current().as_usize(), self.tid);
+
+        let mut value = Some(value);
+
+        // Can we fit the value into an existing page?
+        for page in self.pages.iter_mut() {
+            if let Some(poff) = page.insert(&mut value) {
+                self.len.fetch_add(1, Ordering::Relaxed);
```

Comment:
Using `Relaxed` is highly suspicious to me, but I also recommended removing the counter from the slab above.